img2img instruct-pix2pix support

Klace opened the pull request:
Allows instruct-pix2pix models to be used properly in the img2img pipeline (includes inpainting/masking support). 

**Additional notes**

Changes do not effect the results of txt2img or regular img2img.

Instruct-pix2pix models will use separate CFGDenoiser so there is no impact to txt2img or img2img performance.

Latent Upscale does not work with instruct-pix2pix models (but remains working for regular img2img)

"Image CFG Scale" from original paper demo has been added with a range of 0->3. Set to 1.0 is the equivalent of not using the Image CFG Scale at all and it will use original CFGDenoiser instead. Values between 1.0 and 1.5 are generally good.

AND only works if "Image CFG Scale" = 1. In the future maybe AND can work with different values.

Denoising strength works as expected, usually best results between are 0.75 and 1.0. 

Works best with 512x512 images. Upscaling your output requires more CFG and weight for effects to be applied.



**Environment this was tested in**

 - OS: Windows
 - Browser: Chrome
 - Graphics card: RTX3090

![image](https://user-images.githubusercontent.com/26013475/216742700-e289e90b-0cdc-4cdf-9ea5-222f7b4dda38.png)

Inpainting allows high Denoising and CFG without affecting rest of the picture:

![image](https://user-images.githubusercontent.com/26013475/216737036-21205dfd-57f5-4461-96fb-b84c3fc95b32.png)

Klace committed:
instruct-pix2pix support

Klace committed:
Revert "instruct-pix2pix support"

This reverts commit 269833067de1e7d0b6a6bd65724743d6b88a133f.

Klace committed:
Processing only, no CFGDenoiser change

Allows instruct-pix2pix

Klace committed:
Image CFG Added (Full Implementation)

Uses separate denoiser for edit (instruct-pix2pix) models

No impact to txt2img or regular img2img

"Image CFG Scale" will only apply to instruct-pix2pix models and metadata will only be added if using such model

mezotaken commented:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/7481/commits/6c6c6636bb123d664999c888cda47a1f8bad635b
Breaks HRfix

Klace commented:
> [6c6c663](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/6c6c6636bb123d664999c888cda47a1f8bad635b) Breaks HRfix

Yep sorry fix incoming

Klace committed:
txt2img Hires Fix

Klace committed:
Use original CFGDenoiser if image_cfg_scale = 1

If image_cfg_scale is =1 then the original image is not used for the output. We can then use the original CFGDenoiser to get the same result to support AND functionality.

Maybe in the future AND can be supported with "Image CFG Scale"

Jonseed commented:
Nice!! Well done!

Jonseed commented:
@Klace I'm getting an error completing the request when I go to generate. Any ideas what might be wrong?

```
TypeError: cat() received an invalid combination of arguments - got (list, Tensor), but expected one of:
* (tuple of Tensors tensors, int dim, *, Tensor out)
* (tuple of Tensors tensors, name dim, *, Tensor out)
```

Klace commented:
> @Klace I'm getting an error completing the request when I go to generate. Any ideas what might be wrong?
> 
> ```
> File "F:\repos\auto111\stable-diffusion-webui\modules\sd_samplers_kdiffusion.py", line 133, in forward
>     c_crossattn = torch.cat([tensor[a:b]], uncond)
> TypeError: cat() received an invalid combination of arguments - got (list, Tensor), but expected one of:
>  * (tuple of Tensors tensors, int dim, *, Tensor out)
>  * (tuple of Tensors tensors, name dim, *, Tensor out)
> ```

I'm unable to reproduce it on the latest commit. Is this happening just with instruct-pix2pix checkpoints? What settings?

Jonseed commented:
I'm on the latest commit 3e0f9a75. Yes, just instruct-pix2pix checkpoints. I tried different settings, and got the same error. I'll keep testing. Someone on Reddit [said](https://www.reddit.com/r/StableDiffusion/comments/10tjzmf/comment/j77n76i/?utm_source=share&utm_medium=web2x&context=3) they also saw this error, and it was recommended to delete the venv folder and have the server recreate the environment, and that seemed to fix it for them. But I don't really want to do that, if possible.

Jonseed commented:
There appears to be a problem with the negative prompt field. If I shorten the negative prompt, it works. If it is too long, then I get the error. The length indicator of the negative prompt doesn't seem to be working either; could be related...

Jonseed commented:
Looks like any more than 75 tokens in the negative prompt field, and it breaks and gives the error.

Ratinod commented:
Just for clarification. "Image CFG Scale" only works if "Euler a" is selected? If you select "DDIM", then "Image CFG Scale" does not affect anything.

Klace commented:
> Looks like any more than 75 tokens in the negative prompt field, and it breaks and gives the error.

Yep. Thought longer prompts were working but that must have been before the Image CFG was finished. It will work with ImageCFG=1 for now with the original denoiser.

I'll try to fix it along with AND.

wmooring commented:
I'm getting the same tuple/tensors errors.  I got around it by shortening my negative prompt, but now while I can generated an image with it, it's basically no different than the original picure.
